crates/sui-core/src/db_checkpoint_handler.rs

```diff
@@ -0,0 +1,381 @@
+    output_object_store: Arc<DynObjectStore>,
+    /// Time interval to check for presence of new db checkpoint
+    interval: Duration,
+    /// File markers which signal that local db checkpoint can be garbage collected
+    gc_markers: Vec<String>,
+}
+
+impl DBCheckpointHandler {
+    pub fn new(
+        input_path: &std::path::Path,
+        output_object_store_config: &ObjectStoreConfig,
+        interval_s: u64,
+    ) -> Result<Self> {
+        let input_store_config = ObjectStoreConfig {
+            object_store: Some(ObjectStoreType::File),
+            directory: Some(input_path.to_path_buf()),
+            ..Default::default()
+        };
+        Ok(DBCheckpointHandler {
+            input_object_store: input_store_config.make()?,
+            output_object_store: output_object_store_config.make()?,
+            interval: Duration::from_secs(interval_s),
+            gc_markers: vec![UPLOAD_COMPLETED_MARKER.to_string()],
+        })
+    }
+    pub fn new_for_test(
+        input_object_store_config: &ObjectStoreConfig,
+        output_object_store_config: &ObjectStoreConfig,
+        interval_s: u64,
+    ) -> Result<Self> {
+        Ok(DBCheckpointHandler {
+            input_object_store: input_object_store_config.make()?,
+            output_object_store: output_object_store_config.make()?,
+            interval: Duration::from_secs(interval_s),
+            gc_markers: vec![UPLOAD_COMPLETED_MARKER.to_string(), TEST_MARKER.to_string()],
+        })
+    }
+    pub fn start(self) -> Sender<()> {
+        let (sender, mut recv) = channel::<()>();
+        let mut interval = tokio::time::interval(self.interval);
+        tokio::task::spawn(async move {
+            info!("DB checkpoint handler loop started");
+            loop {
+                tokio::select! {
+                    _now = interval.tick() => {
+                        if let Err(err) = self.upload_db_checkpoint_to_object_store().await {
+                            error!("Failed to upload db checkpoint to remote store with err: {:?}", err);
+                        }
+                    },
+                    _ = &mut recv => break,
+                }
+            }
+        });
+        sender
+    }
+    async fn upload_db_checkpoint_to_object_store(&self) -> Result<()> {
+        let local_checkpoints_by_epoch = self
+            .read_checkpoint_dir(self.input_object_store.clone())
+            .await?;
+        let remote_checkpoints_by_epoch = self
+            .read_checkpoint_dir(self.output_object_store.clone())
+            .await?;
+
+        let next_epoch = if let Some((last_epoch, path)) =
+            remote_checkpoints_by_epoch.last_key_value()
+        {
+            let success_marker = path.child(SUCCESS_MARKER);
+            let get_result = self.output_object_store.get(&success_marker).await;
+            match get_result {
+                Ok(_) => last_epoch + 1,
+                Err(Error::NotFound { .. }) => {
```

Comment:
i'm confused by this - the comments below indicate that we should delete the local dir if UPLOAD_COMPLETE_MARKER has been written, but this deletes if SUCCESS_MARKER is not found?

Reply:
ok, i get it now - but why do we need to delete the dir before continuing? shouldn't we assume we can resume where we left off?

Reply:
My main worry was one of the files being copied is only partly copied and application crashed/stopped. In which case on resuming copy we will skip over that file. I am also not sure if the object store api for different cloud vendors will throw an error if a bucket key already exists when we resume copy and that is why found it safe to delete those files first. Overall found it simple to delete and copy as a first pass. Happy to incorporate any suggestions.

Reply:
Some of the object store implementations use multi part upload where they split the underlying data into smaller chunks. So I was not sure, if a file upload is an atomic operation on the cloud store or not (so to be safe I assumed it is not)